README
Object-InsideOut version 3.98
=============================

This module provides comprehensive support for implementing classes using the
inside-out object model.

This module implements inside-out objects as anonymous scalar references that
are blessed into a class with the scalar containing the ID for the object
(usually a sequence number).  Object data (i.e., fields) are stored within the
class's package in either arrays indexed by the object's ID, or hashes keyed
to the object's ID.

The virtues of the inside-out object model over the 'blessed hash' object
model have been extolled in detail elsewhere.  Briefly, inside-out objects
offer the following advantages over 'blessed hash' objects:

* Encapsulation

Object data is enclosed within the class's code and is accessible only through
the class-defined interface.

* Field Name Collision Avoidance

Inheritance using 'blessed hash' classes can lead to conflicts if any classes
use the same name for a field (i.e., hash key).  Inside-out objects are immune
to this problem because object data is stored inside each class's package, and
not in the object itself.

* Compile-time Name Checking

A common error with 'blessed hash' classes is the misspelling of field names:

    $obj->{'coment'} = 'No comment';   # Should be 'comment' not 'coment'

As there is no compile-time checking on hash keys, such errors do not usually
manifest themselves until runtime.

With inside-out objects, data is accessed using methods, the names of which
are checked by the Perl compiler such that any typos are easily caught using
"perl -c".


This module offers all the capabilities of other inside-out object modules
with the following additional key advantages:

* Speed

When using arrays for storing object data, Object::InsideOut objects are as
much as 40% faster than 'blessed hash' objects for fetching and setting data,
and even with hashes they are still several percent faster than 'blessed hash'
objects.

* Threads

Object::InsideOut is thread safe, and thoroughly supports sharing objects
between threads using threads::shared.

* Flexibility

Allows control over object ID specification, accessor naming, parameter name
matching, and more.

* 'Runtime' Support

Supports classes that may be loaded at runtime (i.e., using "eval { require
...; };").  This makes it usable from within mod_perl, as well.  Also supports
additions to class hierarchies, and dynamic creation of object fields during
runtime.

* Perl 5.6

Tested on Perl v5.6.0 through v5.6.2, v5.8.0 through v5.8.8, and v5.9.3.

* Exception Objects

As recommended in "Perl Best Practices", Object::InsideOut uses
Exception::Class for handling errors in an OO-compatible manner.

* Object Serialization

Object::InsideOut has built-in support for object dumping and reloading that
can be accomplished in either an automated fashion or through the use of
class-supplied subroutines.  Serialization using 'Storable' is also supported.

* Foreign Class Inheritance

Object::InsideOut allows classes to inherit from foreign (i.e.,
non-Object::InsideOut) classes, thus allowing you to sub-class other Perl
class, and access their methods from your own objects.

* Introspection

Obtain constructor parameters and method metadata for Object::InsideOut
classes.

INSTALLATION

To install this module type the following:

    perl Makefile.PL
    make
    make test
    make install

or if you have Module::Build installed:

    perl Build.PL
    perl Build
    perl Build test
    perl Build install

DEPENDENCIES

Requires Perl 5.6.0 or later.

This module uses the following 'standard' modules:

  ExtUtils::MakeMaker          - For installation
  Test::More (0.50 or later)   - For installation
  Scalar::Util (1.10 or later) - Standard in 5.8 or obtain from CPAN
  Data::Dumper
  attributes
  overload
  B

This module requires the following module available from CPAN:

  Exception::Class (1.22 or later)

Using the :lvalue accessor feature of this module requires the following
module from CPAN:

  Want (0.12 or later)

For :SECURE mode, this module requires the following module from CPAN:

  Math::Random::MT::Auto (5.04 or later)

COPYRIGHT AND LICENCE

Copyright 2005 - 2012 Jerry D. Hedden <jdhedden AT cpan DOT org>

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

# EOF


Changes
Revision history for Perl extension Object::InsideOut.

3.98 Thu Oct  3 19:04:26 2013
	- User string compare for version checks

3.97 Thu Nov 15 13:34:15 2012
	- Fix for panic during destroy from Krzysztof Lewicki

3.96 Mon Oct  1 12:22:50 2012
	- Tweaks in tests for changes in core warning messages

3.95 Tue Jul 24 13:30:57 2012
	- Delete on arrays is deprecated (removed last vestage)

3.94 Wed May  9 17:29:23 EDT 2012
	- Delete on arrays is deprecated

3.93 Mon Apr  9 13:45:35 2012
	- Allow :Handle to work with non-OIO classes per contribution by Damian Conway

3.92 Tue Mar  6 14:42:27 2012
	- Added readonly fields per contribution by Damian Conway

3.91 Wed Feb 22 16:35:09 2012
	- Added sequential defaults per contribution by Damian Conway
	- Extended delegator capabilities per contribution by Damian Conway

3.89 Thu Feb 16 19:08:31 2012
	- Added generated defaults per contribution by Damian Conway

3.88 Thu Jan 26 14:56:59 2012
	- Update build prereqs

3.87 Thu Jan 19 13:46:51 2012
	- Added missing test file for delegators

3.86 Thu Jan 19 04:37:33 2012
	- Added delegators per contribution by Damian Conway

3.85 Wed Jan 11 06:01:11 2012
	- Fix some 'used only once' warnings

3.84 Tue Oct 11 22:42:26 2011
	- Fixed bug in string dump of objects

3.83 Thu Oct  6 23:53:04 2011
	- Fix bug with NOT_SHARED in threads::shared app
	- Document error caused by returning non-shared objects from threads

3.82 Tue Sep 27 23:28:33 2011
	- Fix obj ID reuse for shared objects in threaded app
	- Documented limitation regarding removing shared internal objects

3.81 Tue Mar 22 02:20:49 2011
	- Faster initialization (from O(n^3) to O(n))

3.79 Fri Dec 24 18:06:02 2010
	- Bundle update

3.78 Fri Dec 24 15:35:25 2010
	- POD update

3.77 Tue Dec 21 14:29:37 2010
	- Test fixes for $@ changes
	- Test change for ??? syntax error

3.76 Mon Nov 15 00:01:50 2010
	- Skip some more shared tests under 5.8.8 and prior

3.75 Tue Nov  9 15:01:06 2010
	- Skip some shared tests under 5.8.8 and prior

3.74 Tue Nov  9 02:16:45 2010
	- threads::shared version number in test

3.73 Mon Nov  8 18:53:00 2010
	- Fix obj ID handling for shared objects in threaded app

3.72 Mon Oct 18 21:15:53 2010
	- Fix obj ID handling for non-shared objects in threaded app

3.71 Wed Oct 13 18:25:20 2010
	- Fixed a memory leak

3.69 Tue Sep 28 20:38:53 2010
	- Fix test counts in t/27-exception.t

3.68 Thu Sep 23 17:09:21 2010
	- POD note to 'use threads' in DBI applications [bug #58394]
	- Allow warnings for unhandled args

3.67 Mon May 17 23:22:44 2010
	- More fixes for $@ changes

3.66 Fri May 14 13:31:04 2010
	- More fixes for $@ changes

3.65 Fri May  7 19:05:34 2010
	- Fix tests for $@ changes that will be coming with Perl 5.14

3.64 Mon Mar  8 17:59:51 2010
	- Added tests related to shared object destruction

3.63 Fri Mar  5 13:11:48 EST 2010
	- Fix for shared object destruction bug

3.62 Thu Mar  4 04:24:02 2010
	- Extended the change in v3.59 to Perl 5.8.9

3.61 Wed Mar  3 21:37:19 2010
	- Code cleanup

3.59 Wed Mar  3 19:11:50 2010
	- Fixed shared object destruction when using Thread::Queue

3.58 Wed Dec 23 12:45:55 2009
	- Support type checking for scalar refs

3.57 Thu Nov  5 01:42:26 2009
	- Check args to ->isa()/->can()

3.56 Wed Jul 29 19:11:02 2009
	- Changed label that was a keyword

3.55 Fri May  8 11:38:53 2009
	- Upgraded dependencies to Exception::Class 1.29

3.54 Wed Feb 18 18:04:51 2009
	- Doc update only

3.53 Wed Feb 18 15:34:20 2009
	- Handle unnamed fields when using Storable
	- Cache some object initialization activities

3.52 Mon Oct 27 11:16:19 2008
	- Upgraded dependencies to Exception::Class 1.26

3.51 Thu Oct 23 20:18:23 2008
	- Fix issues with Exception::Class 1.25

3.49 Fri Oct 17 18:33:46 2008
	- Call all :Destroy methods before deleting field data

3.48 Wed Oct 15 20:05:20 2008
	- Fix runtime loading of classes with multiple inheritance

3.47 Mon Oct  6 18:50:26 2008
	- Proper implementation of ->isa()/->can()

3.46 Fri Oct  3 21:01:59 2008
	- Support Test::MockObject

3.45 Fri Jul 11 18:58:46 2008
	- Don't clone deep internally held objects

3.44 Fri Jul 11 17:39:35 2008
	- Don't clone internally held objects

3.43 Tue Jun 24 18:05:10 2008
	- Added 'scalar' type checking

3.42 Fri Jun 13 17:35:27 2008
	- End all tests with exit(0) and fix SKIPs
	- Recommended modules in META.yml

3.41 Fri May 16 11:16:30 EDT 2008
	- Changed function shared_clone() to clone_shared() in Util.pm
	- Handle circular refs when cloning data

3.39 Mon Mar 10 20:25:30 2008
	- Catch unhandled params to ->new()

3.38 Tue Feb 26 16:54:46 2008
	- No shared copying if no threads

3.37 Tue Feb 19 19:01:56 2008
	- Fix another preformance issue with reclaiming object IDs
	- Fix testing under Perl 5.8.0

3.36 Fri Feb 15 19:45:37 2008
	- Shared copies of refs of refs

3.35 Fri Dec 21 15:18:29 2007
	- Set min threads::shared version when overloading '==' operator
	    for shared objects
	- Added a countdown indicator to Term::YAPI

3.34 Wed Nov 28 01:47:52 2007
	- Fix preformance issue with reclaiming object IDs

3.33 Fri Nov  9 13:16:56 2007
	- Fix overload tests

3.32 Thu Nov  8 21:57:28 2007
	- Skip overload test file if no threads

3.31 Thu Nov  8 19:22:42 2007
	- Overload the '==' operator for shared objects

3.29 Wed Nov  7 18:20:56 2007
	- Fix to dump (Bug #30527)
	- Additional fix to UNIVERSAL::isa call (Bug #30533)

3.28 Wed Nov  7 01:34:02 2007
	- Fix to UNIVERSAL::isa call (Bug #30533)

3.27 Sat Nov  3 01:23:30 2007
	- Make :Default({}) work as expected
	- Added ->endtime() method to Term::YAPI

3.26 Mon Sep 17 23:59:28 2007
	- Support ANSI color sequences in Term::YAPI

3.25 Tue Sep 11 17:29:00 2007
	- Capability to install Term::YAPI (fixed)

3.24 Tue Sep 11 17:07:07 2007
	- Capability to install Term::YAPI

3.23 Tue Sep 11 16:17:44 2007
	- Localize error vars in DESTROY

3.22 Fri Sep  7 19:34:23 2007
	- Enhancements to Term::YAPI in examples dir
	- POD spelling test (maintainer only)

3.21 Fri Aug 17 15:46:35 2007

Build.PL
# Module makefile for Object::InsideOut (using Module::Build)

use strict;
use warnings;

use Module::Build;


# Check for Scalar::Util::weaken()
eval { require Scalar::Util; };
if ($@) {
    # Not found - require minimum version
    $Scalar::Util::VERSION = 1.10;
} elsif (! Scalar::Util->can('weaken')) {
    # Pure Perl version only
    if ($Scalar::Util::VERSION < 1.10) {
        # Require minimum version
        $Scalar::Util::VERSION = 1.10;
    } else {
        # Require next higher version
        $Scalar::Util::VERSION += 0.01;
    }
}


my %more_recommends = ();
if ($] >= 5.008) {
    %more_recommends = (
        'threads'             => 1.42,
        'threads::shared'     => 1.03,
    );
}

# Create make file
Module::Build->new(
    'module_name' => 'Object::InsideOut',
    'license'     => 'perl',
    'dist_author' => 'Jerry D. Hedden <jdhedden AT cpan DOT org>',

    'build_requires' => {
                            'Test::More'          => 0.50,
                        },

    'requires'       => {
                            'perl'                => '5.6.0',
                            'strict'              => 0,
                            'warnings'            => 0,
                            'attributes'          => 0,
                            'overload'            => 0,
                            'Config'              => 0,
                            'B'                   => 0,
                            'Data::Dumper'        => 0,
                            'Scalar::Util'        => $Scalar::Util::VERSION,
                            'Exception::Class'    => 1.22,
                        },

    'recommends'     => {
                            'Want'                => 0.12,
                            'Storable'            => 2.15,
                            'Test::Pod'           => 1.26,
                            'Test::Pod::Coverage' => 1.08,
                            %more_recommends
                        },
)->create_build_script();

# EOF
